src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/bucketize_ref.cl

```diff
@@ -0,0 +1,70 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+// If one input type is float and another is half, we will cast to a smaller type - half
+// This is needed because we cannot correctly compare float and half types without casting
+// Alternative way is to compare bucket value with subtraction of input value and some epsilon
+#define TO_SMALLER_TYPE(X) (X)
+#if INPUT0_IS_FP && INPUT1_IS_FP
+#    if INPUT0_TYPE_SIZE < INPUT1_TYPE_SIZE
+#        define TO_SMALLER_TYPE(X) TO_INPUT0_TYPE(X)
+#    elif INPUT0_TYPE_SIZE > INPUT1_TYPE_SIZE
+#        define TO_SMALLER_TYPE(X) TO_INPUT1_TYPE(X)
+#    endif
+#endif
+
+#define GET_INDEX(prefix, ORDER) CAT(prefix, _GET_INDEX)(ORDER)
+
+KERNEL(bucketize_ref)
+(const __global INPUT0_TYPE* input, const __global INPUT1_TYPE* buckets, __global OUTPUT_TYPE* output) {
+    const uint dim0 = get_global_id(0);
+    const uint dim1 = get_global_id(1);
+    const uint dim2 = get_global_id(2);
+
+#if OUTPUT_DIMS == 4
+#    define ORDER b, f, y, x
+    const uint x = dim0;
+    const uint y = dim1;
+#elif OUTPUT_DIMS == 5
+#    define ORDER b, f, z, y, x
+    const uint x = dim0 % OUTPUT_SIZE_X;
+    const uint y = dim0 / OUTPUT_SIZE_X;
+    const uint z = dim1;
+#elif OUTPUT_DIMS == 6
+#    define ORDER b, f, w, z, y, x
+    const uint x = dim0 % OUTPUT_SIZE_X;
+    const uint y = dim0 / OUTPUT_SIZE_X;
+    const uint z = dim1 % OUTPUT_SIZE_Z;
+    const uint w = dim1 / OUTPUT_SIZE_Z;
+#endif
+    const uint f = dim2 % OUTPUT_FEATURE_NUM;
+    const uint b = dim2 / OUTPUT_FEATURE_NUM;
+
+    const uint index = GET_INDEX(INPUT0, ORDER);
+    const INPUT0_TYPE value = input[index];
+
+    // TODO Maybe take out this binary search to some common utils ?
+    uint first = 0;
+    uint length = INPUT1_LENGTH;
+    while (length > 0) {
+        const uint half_length = length / 2;
+        const uint middle = first + half_length;
+#if WITH_RIGHT_BOUND
+        if (TO_SMALLER_TYPE(buckets[middle]) < TO_SMALLER_TYPE(value)) {
+#else
+        if (TO_SMALLER_TYPE(buckets[middle]) <= TO_SMALLER_TYPE(value)) {
```

Comment:
in lower_bound it's usually >=, are you sure it works correctly for both cases?

Reply:
At least at the current implementation in my machine, std::lower_bound and std::upper_bound both have almost the same logic, and the only difference is that they do if operands are equal: std::lower_bound (with_right_bound in our case) decreases the "length" and don't increase "first", and std::upper_bound (with_left_bound in our case) decreases the "length" and increase "first". Tests check this case, should work correctly.

Reply:
I see, it works because there is no equal values in buckets

Reply:
Yeah, there all should be sorted and unique.